### DISCUSSION TITLE
How to add async/await inside the listen events

### DISCUSSION THREAD

**Discussion by sdicker8** [tauri-apps/tauri#4959] [Q&A] [answered]
I would like to close my database when the app closes, but can't quit figure out how to do it - here's what  i've got. This compiles but panics at runtime : 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.'

        .on_window_event(|event| match event.event() {
            WindowEvent::CloseRequested { api: _, .. } => {
                if let Some(db) = event.window().try_state::<Arc<Pool<Sqlite>>>() {
                    //tokio::runtime::Handle::current().block_on(async {
                    //tauri::async_runtime::handle().block_on(async {
                    tauri::async_runtime::block_on(async {
                        db.close().await;
                    })
                }
            }
            _ => {}
        })

this question was essentially asked here (but there was no final resolution):
https://users.rust-lang.org/t/how-to-add-async-await-inside-the-listen-events-in-tauri-application/78608

Any advice would be greatly appreciated - thanks -

**Comment by FabianLars** [comment, marked as answer]
The last messages in that thread go somewhat in the right direction. 1) It's actually quite rare that you really need tokio::main _in a tauri app. 2) if you still want to do it (which is fine btw), you need to give Tauri a handle to the tokio runtime: https://docs.rs/tauri/latest/tauri/async_runtime/fn.set.html

Edit: It gets a little weird inside the on_X_event handlers, you need to wrap the async_runtime::block_on call in `tauri::task::block_in_place` :/

**Comment by sdicker8** [comment]
FabianLars figured out a solution here - wrap the block_on call in tokio::task::block_in_place
This only necessary if main is an async tokio::main
